Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DataSynchronizerError.swift

```diff
@@ -1,9 +1,65 @@
-public class DataSynchronizerError: Error, CustomStringConvertible {
-    public var description: String
+/// A set of errors that can occur during the synchronization process.
+public enum DataSynchronizerError: Error, CustomStringConvertible {
+    /**
+        Any error related to decoding a document
+        This will likely be due to a malformed sync version
+        or an unsupported change event.
+    */
+    case decodingError(_ message: String)
+    /**
+        An unsupported sync protocol version has been
+        supplied with a document. The error message
+        contains the faulty version.
+    */
+    case unsupportedProtocolVersion(_ message: String)
+    /**
+        Any error related to trying to operate
+        on a document that does not exist.
+        This can occur if a change event is sent to us
+        with missing data, or if a a local document
+        is deleted during a replace or update.
+     */
+    case documentDoesNotExist(_ message: String)
+    /**
+        Any unexpected mongoDB error. Certain errors
+        during the syncing process are expected and will
+        create conflicts (e.g., inserting on the same _id)
+        but errors outside of the scope of what we can handle
+        will be thrown here
+    */
+    case mongoDBError(_ message: String, _ error: Error)
```

Comment:
does this capture the error category from newer versions of mongo (e.g. ability to see if this was a transient network error, etc)?

Reply:
This captures whatever errors could be emitted by MobileMongo.